### ISSUE TITLE
GemNotFoundException

### ISSUE THREAD

**Issue by tsvallender** [Shopify/ruby-lsp#1158]
### Ruby version

3.2.0

### Description

I’m trying to set up ruby-lsp for the first time. I’m using Neovim and have copied the setup code from [the documentation](https://github.com/Shopify/ruby-lsp/blob/main/EDITORS.md). My development environment is [devenv](https://devenv.sh) on NixOS, which is why the paths in the messages below are non-standard.

In Neovim, I get the error message `Client 1 quit with exit code 1 and signal 0`.

In `lsp.log` I get the below:
```
[START][2023-10-31 16:21:05] LSP logging initiated
[ERROR][2023-10-31 16:21:05] .../vim/lsp/rpc.lua:734    "rpc"   "/home/tsv/foxsoft/oxed/zeus/.devenv/state/.bundle/ruby/3.2.0/bin/ruby-lsp"     "stderr"        "/nix/store/n25jzg8gf8lxrwmvc7293a2v78frj0ph-ruby-3.2.2/lib/ruby/3.2.0/rubygems.rb:263:in `find_spec_for_exe': can't find gem ruby-lsp (>= 0.a) with executable ruby-lsp (Gem::GemNotFoundException)\n\tfrom /nix/store/n25jzg8gf8lxrwmvc7293a2v78frj0ph-ruby-3.2.2/lib/ruby/3.2.0/rubygems.rb:282:in `activate_bin_path'\n\tfrom /home/tsv/foxsoft/oxed/zeus/.devenv/state/.bundle/ruby/3.2.0/bin/ruby-lsp:25:in `<main>'\n"
```

I’m struggling to work out exactly what this message is saying doesn’t exist: the ruby-lsp gem certainly does.

### Expected output

Neovim should successfully start and connect to ruby-lsp.

**Comment by tsvallender** [comment]
Possibly worth noting that if I run `ruby-lsp` from the command line it runs through and successfully installs gems, then hangs at `Starting Ruby LSP...` which I understand is expected.

**Comment by vinistock** [comment]
Thank you for the report! The message is indeed saying that it couldn't find the `ruby-lsp` and its executable.

I don't use NeoVim and those instructions are contributed by the community, so not sure what the problem is exactly. It's likely that something is activated in the terminal, but isn't inside NeoVim.

Are you using some sort of Ruby version manager? Is the directory where the `ruby-lsp` executable exists inside the `PATH` from inside NeoVim?

**Comment by tsvallender** [comment]
Thanks for the feedback. Hm, I guess that does make sense—devenv/Nix manage the Ruby version, and despite starting Neovim from a shell with the correct PATH I wonder if it ends up running `ruby-lsp` from an environment _without_ the correct PATH. I think I did try installing ruby-lsp globally without further luck—does it then still re-install in the local environment which would again cause it to fail? I'll try that approach again shortly and see.

**Comment by vinistock** [comment]
It might be helpful to explain what we do in the VS Code extension. We start by activating the correct Ruby version using whatever version manager the user picked (chruby, rbenv, asdf...). That will insert the right things inside `PATH`, `GEM_HOME` and all other relevant environment variables.

Then we install the `ruby-lsp` gem globally, so that we can run the custom bundle logic (literally just `gem install ruby-lsp`). When you run `ruby-lsp` (without bundle exec), we create a custom bundle under the `.ruby-lsp` directory, which is a combination of the project's bundle + the `ruby-lsp` and `debug`.

After running `bundle install` and finishing the custom bundle setup, we use `exec` to replace the current process by running `BUNDLE_GEMFILE=.ruby-lsp/Gemfile bundle exec ruby-lsp`, which then boots the server with your project's gems inside the `$LOAD_PATH` automatically.

Long story short, as long as the right Ruby version is activated and you can run `ruby-lsp`, then it should work.

**Comment by tsvallender** [comment]
Thanks for the explanation. I did some more digging and fiddling and have got it connecting, though I tried so many different things I'm not sure what actually stuck and I'm still getting some errors around Rubocop. I'll try and unpick it later and document the working setup if I can.
